src/enzo/FofLib.C

```diff
@@ -0,0 +1,1477 @@
+      SearchVarTree(link, x, x+3*fifo[fifohead], link[fifo[fifohead]],
+		    root, fifo, &fifotail, group, groupnum,
+		    *groupsize+groupnum);
+    groupnum++; /* Increment group number */
+  }
+  
+  /* Free up unneeded memory */
+  free(fifo);
+  *groupsize=(int *) realloc(*groupsize, groupnum*sizeof(int));
+  free(idxlist);
+  free(root+1);
+
+  return(groupnum); /* Return number of groups */
+}
+
+int FofVarList(int npart, FLOAT *x, FLOAT *link, int *group, int
+	**groupsize, int ***grouplist) {
+  int fifohead, fifotail, *idxlist;
+  int groupnum=0;
+  int n, m;
+  struct treenode *root;
+  
+  /* First allocate workspace and maximum size for output array
+     groupsize */
+  if (!(idxlist=(int *) calloc(npart,sizeof(int))))
+    ErrorHandler("unable to allocate workspace in FofList");
+  if (!(*groupsize=(int *) calloc(npart,sizeof(int))))
+    ErrorHandler("unable to allocate workspace in FofList");
+  if (!(*grouplist=(int **) calloc(npart,sizeof(int *))))
+    ErrorHandler("unable to allocate workspace in FofList");
+  
+  /* Initialize group list array */
+  for (n=0; n<npart; n++) {
+    group[n]=-1;
+    idxlist[n]=n;
+  }
+  
+  /* Build tree of particle positions */
+  root=BuildVarTree(x, link, npart, idxlist, LEAFSIZE);
+
+  /* Now proceed through particle list */
+  for (n=0; n<npart; n++) {
+    if (group[n]!=-1) continue; /* Already in a group ? */
+    else group[n]=groupnum; /* Assign groupnum to particle */
+    /* Allocate space for particle list */
+    if (!((*grouplist)[groupnum]=(int *) calloc(npart,sizeof(int))))
+      ErrorHandler("unable to allocate workspace in FofList");
+    (*groupsize)[groupnum]=1; /* Initial size of group */
+    (*grouplist)[groupnum][0]=n; /* Load first particle into fifo */
+    /* Start with fifo head and tail pointers both at 0. In each step,
+       find unassigned particles within linking length of particle
+       whose index is fifo[fifohead]. Add them at the end of fifo and
+       increment fifotail, because they need to be checked too. Once
+       fifohead passes fifotail, we've checked all particles in this
+       group and we're done. */
+    for (fifohead=fifotail=0; fifohead<=fifotail; fifohead++)
+      /* Search for friends and add to fifo */
+      SearchVarTree(link, x, x+3*(*grouplist)[groupnum][fifohead],
+		 link[(*grouplist)[groupnum][fifohead]], root,
+		 (*grouplist)[groupnum], &fifotail, group, groupnum,
+		 *groupsize+groupnum);
+    /* Free unused parts of group list */
+    (*grouplist)[groupnum]=(int *) realloc((*grouplist)[groupnum],
+				   fifotail*sizeof(int));
+    groupnum++; /* Increment group number */
+  }
+  
+  /* Free up unneeded memory */
+  *groupsize=(int *) realloc(*groupsize, groupnum*sizeof(int));
+  *grouplist=(int **) realloc(*grouplist, groupnum*sizeof(int *));
+  free(idxlist);
+  free(root+1);
+
+  return(groupnum); /* Return number of groups */
+}
+
+int Fof(int npart, FLOAT *x, FLOAT link, int *group, int **groupsize)
```

Comment:
Is this function used?

Reply:
Yes by the active particle merging function. It is used to find groups for merging.